File: PRACTICA02/JAVA - CLASE FRACCION/Fraccion.java
public class Fraccion {

private int numerador;
private int denominador;

public Fraccion () {
this.numerador = 0;
this.denominador = 1;
}

public Fraccion(int n, int d) {
this.numerador = n;
this.denominador = d;
}

@Override
public boolean equals(Object o) {
if (o instanceof Fraccion) {
Fraccion f = (Fraccion) o;
return f.numerador == this.numerador && f.denominador == this.denominador;
}
return false;
}

@Override
public String toString () {
return String.format("%d/%d", this.numerador , this.denominador);
}

public Fraccion suma(Fraccion o) {
int n = (this.numerador * o.denominador) + (this.denominador * o.numerador);
int d = this.denominador * o.denominador;
return new Fraccion(n, d);
}

public Fraccion resta(Fraccion o) {
Fraccion c = new Fraccion ();
c.numerador = (this.numerador * o.denominador) - (this.denominador * o.numerador);
c.denominador = this.denominador * o.denominador;
return c;
}

public Fraccion multiplica(Fraccion o){
return new Fraccion(this.numerador*o.numerador,this.denominador*o.denominador);
}

public Fraccion divide(Fraccion o){
return new Fraccion(this.numerador*o.denominador,this.denominador*o.numerador);
}

public double convertirADecimal(){
if(this.denominador==0) return 0;
return (double)this.numerador/this.denominador;
}

public boolean esInverso(Fraccion o){
Fraccion r = this.multiplica(o);
return r.numerador == r.denominador;
}

public static Fraccion parseFraccion(String str){
String[] p = str.split("/");
return new Fraccion(Integer.parseInt(p[0].trim()),Integer.parseInt(p[1].trim()));
}

public Fraccion simplifica(){
int a = Math.abs(this.numerador);
int b = Math.abs(this.denominador);
while(b!=0){
int t=b;
b=a%b;
a=t;
}
return new Fraccion(this.numerador/a,this.denominador/a);
}
}

File: PRACTICA02/PYTHON - CLASE FRACCION/fraccion.py
class Fraccion:

    def __init__(self ,a,b):
        self.__numerador = a
        self.__denominador = b

    def __add__(self , o):
        a = self.__numerador * o.__denominador + self.__denominador * o.__numerador
        b = self.__denominador * o.__denominador
        return Fraccion(a,b)

    def __sub__(self , o):
        c = Fraccion (0,1)
        c.__numerador = self.__numerador * o.__denominador - self.__denominador * o.__numerador
        c.__denominador = self.__denominador * o.__denominador
        return c

    def __mul__(self,o):
        return Fraccion(self.__numerador*o.__numerador,self.__denominador*o.__denominador)

    def __truediv__(self,o):
        return Fraccion(self.__numerador*o.__denominador,self.__denominador*o.__numerador)

    def __eq__(self,o):
        return self.__numerador==o.__numerador and self.__denominador==o.__denominador

    def convertirADecimal(self):
        if self.__denominador==0:
            return 0
        return self.__numerador/self.__denominador

    def esInverso(self,o):
        r = self * o
        return r.__numerador == r.__denominador

    def parseFraccion(self,str):
        p = str.split("/")
        return Fraccion(int(p[0]),int(p[1]))

    def simplifica(self):
        a = abs(self.__numerador)
        b = abs(self.__denominador)
        while b!=0:
            t=b
            b=a%b
            a=t
        return Fraccion(self.__numerador//a,self.__denominador//a)

    def __str__(self):
        return "{}/{}".format(self.__numerador ,self.__denominador)
